### PR TITLE
health: Remove spawn_netns.sh

### PR DESCRIPTION
**Pull request by gandro**
This commit removes the `spawn_netns.sh` script, which is used to set up a network namespace for the endpoint connectivity test, and replaces it with custom code written in Go. We however still rely on iproute2 to spawn the child process in the new network namespace, as doing this in pure Go tends to be error-prone.

The logic of the network namespace configuration is overall kept the same. There are however a few changes with regards to cleanup:

  - Instead of moving out the veth peer into the host namespace and deleting the device in the host namespace, we just delete the namespace. The kernel will remove any virtual interfaces in that namespace and therefore also remove the veth pair. We already rely on this for ipvlan slaves since e5b9eb32.

  - In a similar fashion, we also do not manually remove ip addresses on the interfaces, we rely on the kernel to remove them with the interface. This further simplifies the cleanup code.

  - The shell script was responsible to clean up the namespace when it received a termination signal (i.e. due to cilium being shut down). With this change, it is now the cilium-agent which listens for a
    termination signal and performs the cleanup. The cleanup functionality is also still called by the `cilium-health-ep` controller if the child exits unexpectedly.

By removing the bash process previously required for cleanup, this change saves roughly 3.9MB of memory.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8349)

### CODE REVIEW

**Comment by gandro** [comment]
test-me-please

File: bpf/Makefile [hunk 1]
```diff
@@ -6,7 +6,6 @@ SUBDIRS = sockops
 
 BPF_SIMPLE = bpf_netdev.o bpf_overlay.o bpf_xdp.o bpf_ipsec.o bpf_network.o bpf_alignchecker.o bpf_hostdev_ingress.o
 BPF = bpf_lxc.o bpf_lb.o bpf_sock.o $(BPF_SIMPLE)
-SCRIPTS = init.sh join_ep.sh run_probes.sh spawn_netns.sh
```
Comment (gandro) [inline bpf/Makefile:9]: Note: Removed it since this has not been in use since e086c90bb :)

File: cilium-health/launch/endpoint.go [hunk 2]
```diff
@@ -107,6 +108,42 @@ func configureHealthRouting(netns, dev string, addressing *models.NodeAddressing
 	return err
 }
 
+func configureHealthInterface(netNS ns.NetNS, ifName string, ip4Addr, ip6Addr *net.IPNet) error {
+	return netNS.Do(func(_ ns.NetNS) error {
```
Comment (brb) [inline cilium-health/launch/endpoint.go:112]: Did you check whether any function call inside the closure does not spawn a goroutine (ctx: https://www.weave.works/blog/linux-namespaces-golang-followup)?

**Review by brb — COMMENTED** [review]
This is great!

Perhaps you could add a link to the kernel mailing-list discussion about removing a device via a netns removal into the PR description.

File: cilium-health/launch/endpoint.go (hunk 2 above)
Comment (gandro) [reply cilium-health/launch/endpoint.go:112]: Yes, I did check for that and didn't find any suspicious code which spawns goroutines. I based my code on very similar code from `pkg/endpoint/connector`, which also invokes similar `netlink` commands in a `netNS.Do()` callback.

**Comment by gandro** [comment]
@brb 
> Perhaps you could add a link to the kernel mailing-list discussion about removing a device via a netns removal into the PR description.

Yes. So I found this discussion in the containers ML from 2011 (with a withdrawn patch) in which people concluded that user-space can expect virtual interfaces (such as veth) to be removed: https://lists.linuxfoundation.org/pipermail/containers/2011-February/026817.html

The `ip netns delete` manpage also says that _physical_ interfaces will be moved out of the namespace prior to deletion. But it doesn't specify the behaviour for virtual interfaces. Unfortunately, I have not found a more authoritative source. In my testing in the vagrant box though it seems to do what I want.

**Review by brb — APPROVED** [review]

**Review by vadorovsky — APPROVED** [review]

**Pull request merged**